### PR TITLE
user: add randomized tests & refactor indexer

### PR DESCRIPTION
**Pull request by ananas-block**
### Randomized Tests:
- added randomized test in light system programs:
- three modes sol only, spl only, both
- these tests are not intended to run in ci but are only for manual testing
Test setup:
- create a relayer from fresh keypairs
- initialize a number of users
- in an infinite loop do:
  - select random user
  - select random token if both
  - select random action
  - merge utxos in case there are utxos in the inbox
  - execute action
 
### indexRecentTransaction:
- streamlined processIndexedTransaction
  - find the inner transaction which calls one of the system verifiers
  - in the instruction the order of account keys is deterministic
  - thus once we have the instruction we can assign the account keys
  - derive the type and amounts
  - compute change amount (to be consistent with solana explorer)
- adapted the order of account keys in verifiers one, storage, and two to be consistent with verifier zero
- I moved variable accounts which are not part of verifier zero to the bottom of the account struct so that the order of the previous accounts is consistent
  @vadorovsky we need to add the change into the account macro as well.

### CODE REVIEW

**Comment by SwenSchaeferjohann** [comment]
Hypothetically, if we want to index custom PSPs does the system work even if devs change the order of account keys?

**Comment by ananas-block** [comment]
> Hypothetically, if we want to index custom PSPs does the system work even if devs change the order of account keys?

we only look at the instructions which call a system verifier, psp devs cannot change the order of the accounts for those instructions.

File: light-system-programs/tests/user-random.tests.ts [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+  });
+
+  /**
+   * fails at tx 22
+   * - had two shields before one usdc and one sol
+   * - history only finds the lastest sol shield
+   *
+   * Check:
+   * - that usdc transactions are categorized correctly
+   * */
```
Comment (SwenSchaeferjohann) [inline light-system-programs/tests/user-random.tests.ts:469]: I suppose this is wip?

File: light-zk.js/src/constants.ts [hunk 2]
```diff
@@ -61,6 +61,13 @@ export const LOOK_UP_TABLE = new PublicKey(
   "DyZnme4h32E66deCvsAV6pVceVw8s6ucRhNcwoofVCem",
 );
 
+export const MAX_U64 = new anchor.BN("18446744073709551615");
+export const VERIFIER_PUBLIK_KEYS = [
```
Comment (SwenSchaeferjohann) [inline light-zk.js/src/constants.ts:65]: Typo publik

File: light-zk.js/src/transaction/indexRecentTransactions.ts [hunk 3]
```diff
@@ -136,88 +163,101 @@ async function processIndexedTransaction(
...
+    publicAmountSpl: Uint8Array,
+    publicAmountSol: Uint8Array,
+  ) => {
+    let type = Action.SHIELD;
+    let amountSpl = new BN(publicAmountSpl, 32, "be");
+    let amountSol = new BN(publicAmountSol, 32, "be");
+
+    let splIsu64 = amountSpl.lte(MAX_U64);
+    let solIsu64 = amountSol.lte(MAX_U64);
+    if (!splIsu64 || !solIsu64) {
```
Comment (SwenSchaeferjohann) [inline light-zk.js/src/transaction/indexRecentTransactions.ts:198]: Wouldnt it be solIsU64 with capitalised U?

**Review by SwenSchaeferjohann — COMMENTED** [review]
Lgtm would like @vadorovsky taking a look too given the scope of the PR

File: light-zk.js/src/constants.ts (hunk 2 above)
Comment (ananas-block) [reply light-zk.js/src/constants.ts:65]: lol

File: light-system-programs/tests/user-random.tests.ts (hunk 1 above)
Comment (ananas-block) [reply light-system-programs/tests/user-random.tests.ts]: yes, and the tx history is out of scope for this pr

**Review by SwenSchaeferjohann — APPROVED** [review]
lgtm

**Pull request merged**